### ISSUE TITLE
Weird implementation of GPT2

### ISSUE THREAD

**Issue by hyunwoongko** [huggingface/transformers#13970]
https://github.com/huggingface/transformers/blob/master/src/transformers/models/gpt2/modeling_gpt2.py#L210

At this line, the attention mask and the attention weight are added.

`attention_mask` input like this.
```python
tensor([[[[1, 1, 1, 1]]],
        [[[1, 1, 0, 0]]]], device='cuda:0')
```

and example attn_weight is 
```python
tensor([[-4.1797e-01, -1.0000e+04, -1.0000e+04, -1.0000e+04],
        [ 3.3545e-01, -4.2328e-02, -1.0000e+04, -1.0000e+04],
        [ 5.5908e-02,  1.7993e-01,  7.2388e-02, -1.0000e+04],
        [ 3.9307e-01,  3.5571e-01,  1.4648e-01,  5.7275e-01]], device='cuda:0',
       dtype=torch.float16, grad_fn=<SelectBackward>)

tensor([[ 1.1707e-01, -1.0000e+04, -1.0000e+04, -1.0000e+04],
        [-1.6565e-01,  4.2285e-01, -1.0000e+04, -1.0000e+04],
        [-4.5996e-01,  1.1963e-01,  1.8762e-01, -1.0000e+04],
        [-1.6150e-01, -5.0439e-01, -2.0889e-02, -4.1901e-02]], device='cuda:0',
       dtype=torch.float16, grad_fn=<SelectBackward>)
```

finally `attn_weight + attention_mask` is 
```python
tensor([[ 5.8203e-01, -1.0000e+04, -1.0000e+04, -1.0000e+04],
        [ 1.3359e+00,  9.5752e-01, -1.0000e+04, -1.0000e+04],
        [ 1.0557e+00,  1.1797e+00,  1.0723e+00, -1.0000e+04],
        [ 1.3926e+00,  1.3555e+00,  1.1465e+00,  1.5723e+00]], device='cuda:0',
       dtype=torch.float16, grad_fn=<SelectBackward>)

tensor([[ 1.1172e+00, -1.0000e+04, -1.0000e+04, -1.0000e+04],
        [ 8.3447e-01,  1.4229e+00, -1.0000e+04, -1.0000e+04],
        [ 5.4004e-01,  1.1191e+00,  1.8762e-01, -1.0000e+04],
        [ 8.3838e-01,  4.9561e-01, -2.0889e-02, -4.1901e-02]], device='cuda:0',
       dtype=torch.float16, grad_fn=<SelectBackward>)
```

Is this correct? Is this implementation intended?
there are two questions.

1. Rather than adding the value, we need to reduce the part to be masked through `torch.mask_fill` to a value such as `-10000`.
2. It should be reflected in both query and key dimension. Currently, this addition only works for one.

We can create an exact mask with code like this:

```python
def create_pad_mask(attention_mask, key_length):
    # type: (Tensor, int) -> Tensor
    return attention_mask.repeat(1, 1, key_length, 1)
    return (mask + mask.permute(0, 1, 3, 2)) <= 1
```
result:
```python
[[[False, False,  False,  True],
   [False, False,  False,  True],
   [ False,  False,  False,  True],
   [ True,  True,  True,  True]]]], device='cuda:0')
```
